File: examples/school/ts/src/token-hmac.ts
import { createHmac, timingSafeEqual } from "node:crypto";

import { ErrTokenIsInvalid } from "./domain.js";
import type { TokenIssuer } from "./providers.js";

// HmacTokens Fulfils TokenIssuer with node:crypto only.
// Reference: https://nodejs.org/api/crypto.html#cryptocreatehmacalgorithm-key-options
export class HmacTokens implements TokenIssuer {
  constructor(
    private readonly secret: string,
    private readonly lifeMs: number,
    private readonly now: () => number = Date.now,
  ) {}

  // issueAccessToken Signs a Subject together with its Deadline.
  issueAccessToken(subject: string): string {
    const claim = `${subject}.${this.now() + this.lifeMs}`;

    return `${claim}.${this.signTokenClaim(claim)}`;
  }

  // readAccessToken Returns the Subject, or Throws why it Cannot.
  // The Comparison Reads forward: a Token Dies once its Deadline Passes.
  readAccessToken(token: string): string {
    const claim = this.openSignedClaim(token);
    const [subject, deadline] = claim.split(".");

    if (this.now() > Number(deadline)) throw ErrTokenIsInvalid;

    return subject!;
  }

  // openSignedClaim Returns the Claim only when the Signature Holds.
  private openSignedClaim(token: string): string {
    const cut = token.lastIndexOf(".");
    if (cut < 0) throw ErrTokenIsInvalid;

    const claim = token.slice(0, cut);
    const given = Buffer.from(token.slice(cut + 1));
    const wanted = Buffer.from(this.signTokenClaim(claim));

    if (given.length !== wanted.length || !timingSafeEqual(given, wanted)) throw ErrTokenIsInvalid;

    return claim;
  }

  // signTokenClaim Reduces a Claim to its Signature.
  private signTokenClaim(claim: string): string {
    return createHmac("sha256", this.secret).update(claim).digest("base64url");
  }
}

File: examples/school/go/token_hmac.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"strconv"
	"strings"
	"time"
)

// ErrTokenIsInvalid Covers every Way a Token can Fail.
var ErrTokenIsInvalid = errors.New("token is Invalid or Expired")

// HmacTokens Fulfils TokenIssuer with the standard Library only.
// Reference: https://pkg.go.dev/crypto/hmac
type HmacTokens struct {
	Secret []byte
	Life   time.Duration
	Now    func() time.Time
}

// IssueAccessToken Signs a Subject together with its Deadline.
func (h HmacTokens) IssueAccessToken(subject string) (string, error) {
	if len(h.Secret) == 0 {
		return "", ErrTokenIsInvalid
	}

	claim := subject + "." + strconv.FormatInt(h.Now().Add(h.Life).Unix(), 10)

	return claim + "." + h.SignTokenClaim(claim), nil
}

// ReadAccessToken Returns the Subject, or Says why it Cannot.
// The Comparison Reads forward: a Token Dies once its Deadline Passes.
func (h HmacTokens) ReadAccessToken(token string) (string, error) {
	claim, err := h.OpenSignedClaim(token)
	if err != nil {
		return "", err
	}

	subject, deadline, _ := strings.Cut(claim, ".")
	seconds, parseErr := strconv.ParseInt(deadline, 10, 64)
	if parseErr != nil || h.Now().After(time.Unix(seconds, 0)) {
		return "", ErrTokenIsInvalid
	}

	return subject, nil
}

// OpenSignedClaim Returns the Claim only when the Signature Holds.
func (h HmacTokens) OpenSignedClaim(token string) (string, error) {
	cut := strings.LastIndex(token, ".")
	if cut < 0 {
		return "", ErrTokenIsInvalid
	}

	claim, signature := token[:cut], token[cut+1:]

	if !hmac.Equal([]byte(signature), []byte(h.SignTokenClaim(claim))) {
		return "", ErrTokenIsInvalid
	}

	return claim, nil
}

// SignTokenClaim Reduces a Claim to its Signature.
func (h HmacTokens) SignTokenClaim(claim string) string {
	mac := hmac.New(sha256.New, h.Secret)
	mac.Write([]byte(claim))

	return base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
}
